services/meta/client.go

```diff
@@ -399,34 +408,61 @@ func (c *Client) MarshalBinary() ([]byte, error) {
 	return nil, nil
 }
 
-func (c *Client) MetaServers() []string {
-	c.mu.RLock()
-	defer c.mu.RUnlock()
-	return c.metaServers
-}
-
 func (c *Client) index() uint64 {
 	c.mu.RLock()
 	defer c.mu.RUnlock()
 	return c.data.Index
 }
 
-// retryUntilExec will attempt the command on each of the metaservers and return on the first success
+// retryUntilExec will attempt the command on each of the metaservers until it either succeeds or
+// hits the max number of tries
 func (c *Client) retryUntilExec(typ internal.Command_Type, desc *proto.ExtensionDesc, value interface{}) error {
 	var err error
 	var index uint64
-	for _, s := range c.MetaServers() {
-		index, err = c.exec(s, typ, desc, value)
+	tries := 0
+	currentServer := 0
+	var redirectServer string
+
+	for {
+		// build the url to hit the redirect server or the next metaserver
+		var url string
+		if redirectServer != "" {
+			url = redirectServer
+			redirectServer = ""
+		} else {
+			c.mu.RLock()
+			if currentServer >= len(c.metaServers) {
+				currentServer = 0
+			}
+			server := c.metaServers[currentServer]
+			c.mu.RUnlock()
+
+			url = fmt.Sprintf("://%s/execute", server)
+			if c.tls {
+				url = "https" + url
+			} else {
+				url = "http" + url
+			}
+		}
+
+		index, err = c.exec(url, typ, desc, value)
+		tries++
```

Comment:
redirect is cleared above right after it gets set. updating the `currentServer` increment